### PR TITLE
feat: Execution Workflow CSI-UTR

### LINKED ISSUE #31: Execution Workflow: CSI-UTR
## WHAT    
**Write** execution workflow for [CSI-UTR](https://github.com/UofLBioinformatics/CSI-UTR). Use the provided [small files for testing](https://github.com/iRNA-COSI/APAeval/tree/main/tests/test_data) (running the workflow on real data is a different issue).

## CHECKLIST   
* [x] Use [snakemake template](https://github.com/iRNA-COSI/APAeval/tree/main/docs/templates/snakemake) or [nextflow template](https://github.com/iRNA-COSI/APAeval/tree/main/docs/templates/nextflow) to create your workflow.
* [ ] Comment your code
* [x] Run individual rules/processes in either conda envs or docker/singularity containers for reproducibility
* [x] Input:  .bam or .fastq from [test_data]((https://github.com/iRNA-COSI/APAeval/tree/main/tests/test_data))
* [ ] Give [feedback about the method](https://forms.gle/mAAkbRJqhedJnSeJ6)

## OUTPUTS (see [specification](https://github.com/iRNA-COSI/APAeval/blob/main/execution_workflows/execution_output_specification.md)): 
* [ ] Does CSI-UTR perform PAS identification? If yes:
* [ ] Output: Adhere to output specification for Identification challenge
     > This BED file contains single-nucleotide position of poly(A) sites identified by the tool.
Fields:

    > chrom - the name of the chromosome
    chromStart - the starting position of the feature in the chromosome
    chromEnd - the ending position of the feature in the chromosome; as identified PAS are single-nucleotide, the ending position is the same as starting position
    name - defines the name of the identified poly(A) site
    score - not used, leave as "."
    strand - defines the strand; either "." (=no strand) or "+" or "-".

* [ ] Output: Adhere to output specification for quantification challenge
     > This BED file contains positions of unique poly(A) sites with TPM values for each identified site in the score column.

    > chrom - the name of the chromosome
    chromStart - the starting position of the feature in the chromosome
    chromEnd - the ending position of the feature in the chromosome; as identified PAS are single-nucleotide, the ending position is the same as starting position
    name - defines the name of the identified poly(A) site
    score - TPM value for the identified site
    strand - defines the strand; either "." (=no strand) or "+" or "-".

* [ ] Output: Adhere to output specification for differential usage challenge
   > This TSV file contains two columns:

    > - gene ID
    > - significance of differential PAS usage

    > Column names should not be added to the file.

### PR DESCRIPTION
**Pull request by faricazjj**
Fixes #31 

Credits to @mrgazzara for providing additional test data and helping with results interpretation for this tool! :D 

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have updated corresponding READMEs (if applicable)
- [x] My code follows the templates/style guidelines of the repository
- [x] In- and output formats comply with APAeval specifications
- [x] No parameters or file names are hardcoded
- [x] Results, logs or other output is not commited to the repository
- [x] I have tested my code

### CODE REVIEW

File: execution_workflows/CSI-UTR/bin/postprocess_differential.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    with open(output_file, 'r') as f:
+        reader = csv.reader(f, delimiter='\t')
+        # skip the header line
+        next(reader)
+        for line in reader:
+            gene = line[1]
+            pvalue = line[6]
+            output = (gene, pvalue)
+            # write to output file
+            differential_out.write("\t".join(output) + "\n")
```
Comment (SamBryce-Smith) [inline execution_workflows/CSI-UTR/bin/postprocess_differential.py:49]: I see duplicates for genes in the differential output (this feels like a problem we've had to address before...). I think in other workflows we pick the smallest p-value per-gene to report (e.g. if tests multiple diff transcript models of the same gene). Could we also do something similar here?

image

Also not sure about whether version numbers for gene IDs being absent here will be a problem or not (e.g. 'ENSMUSG00000036206.1'). I think for other workflows **where we control the input annotation generation** we don't strip the version numbers, and I assume for ground truth data we keep it as is. I don't have a preference in either direction, it's just we have to be consistent with the ground truth data (or adapt how the summary workflow matches gene IDs) such that the gene IDs will match. I understand that this is out of our control with CSI-UTR as we're currently using the pre-provided annotations, but this still needs to be documented

**Review by SamBryce-Smith — COMMENTED** [review]
Thanks for this @faricazjj! I followed the instructions in the README to the letter and got a successful test run which is great! Have a few comments on the output files below though (some of which need wider discussion).

image

Can't comment on these files directly as the diffs are too large, but it looks like `bin/control_identification_out.bed`, `bin/siControl_R1_2genes_identification_out.bed` & `bin/control_vs_srsf3_differential_out.tsv` are outputs from the workflow and don't need to be included in the repo.

When I checked the identification output files they are all identical.
image 

Seems extremely unlikely you'd see the same output across all samples... I suspect these 'identification' sites come from the input annotation, especially as the lengths of the files are identical?
image

In which case I don't think the tool performs **de-novo polyA site identification** and so should not qualify for the identification challenge. Instead it seems it just looks for differential expression/usage of provided annotated sites. From a quick skim through the paper this seems to be the case, but maybe with closer reading I am wrong.

I'm also against us using author pre-provided annotations as input to our pipelines as it introduces a source of variation to comparisons other than the individual tool's performance. It also goes against how we've designed other workflows (i.e. start with a common input of GTF & BED format annotation files, convert to tool-specific formats internally).

Unfortunately CSI-UTR does not provide commands/source code to generate their provided annotations (from what I can see), so implementing this ourselves would be a daunting task. The manuscript at least provides instructions on how they did this, but specific steps may be open to interpretation. To my memory we have also had to implement our own code to generate annotations for Roar, although I think that's the only one (and it hasn't been merged yet). 

To be clear on the final point RE annotations, I'm raising it because I think it's important but I don't want to push for this to be implemented until we've had a wider discussion and come to a group decision. In an ideal world I'd say we implement it ourselves according the to the steps described in the Methods. However, I'm very aware we're tight for time, and this method likely only qualifies for the differential challenge which we are putting at a lower priority for now. Any output from our implementation would likely deviate considerably from theirs too as we'd use the whole PolyASite database whereas the authors used 3'seq data from a single study (GEO series GSE30198; SRA accession SRP007359 - ['Defining Cleavage Site Intervals (CSIs)' in Methods section of CSI-UTR paper](https://www.frontiersin.org/articles/10.3389/fgene.2019.00182/full#h3))

File: execution_workflows/CSI-UTR/bin/postprocess_differential.py (hunk 1 above)
Comment (dominikburri) [reply execution_workflows/CSI-UTR/bin/postprocess_differential.py:49]: Agree, we definitely need to remove the duplicates. For the test data I get over 50'000 rows in `control_vs_srsf3_differetinal_output.tsv` with some genes appearing 5 times or more. 

As for the version numbers: not sure how to handle this right now, but we can always strip the version number in the summary workflow. I agree, a note in the README that the genes are without version number should be there and I hope that suffices for now.

**Comment by dominikburri** [comment]
> When I checked the identification output files they are all identical. image
> 
> Seems extremely unlikely you'd see the same output across all samples... I suspect these 'identification' sites come from the input annotation, especially as the lengths of the files are identical? image
> 
> In which case I don't think the tool performs **de-novo polyA site identification** and so should not qualify for the identification challenge. Instead it seems it just looks for differential expression/usage of provided annotated sites. From a quick skim through the paper this seems to be the case, but maybe with closer reading I am wrong.

Agree, it seems that the identification_out.bed simply contain the content of `Mm10.CSIs.bed`, but with 1nt windows instead of the variable sizes. It looks like the highest number in column 4 from `Mm10.CSIs.bed` was taken as the start position in identification_out.bed (irrespective of strandedness, which seems faulty to me).
Given that we use their annotations and locations, maybe the identification was just offloaded, was performed on their site and provided for us? But then it more resembles a polyA site database. 
So I lean also towards that the tool does not perform identification and the respective run mode in `modules.config` should always be set to *false*. 

> I'm also against us using author pre-provided annotations as input to our pipelines as it introduces a source of variation to comparisons other than the individual tool's performance. It also goes against how we've designed other workflows (i.e. start with a common input of GTF & BED format annotation files, convert to tool-specific formats internally).
> 
> Unfortunately CSI-UTR does not provide commands/source code to generate their provided annotations (from what I can see), so implementing this ourselves would be a daunting task. The manuscript at least provides instructions on how they did this, but specific steps may be open to interpretation. To my memory we have also had to implement our own code to generate annotations for Roar, although I think that's the only one (and it hasn't been merged yet).
> 
> To be clear on the final point RE annotations, I'm raising it because I think it's important but I don't want to push for this to be implemented until we've had a wider discussion and come to a group decision. In an ideal world I'd say we implement it ourselves according the to the steps described in the Methods. However, I'm very aware we're tight for time, and this method likely only qualifies for the differential challenge which we are putting at a lower priority for now. Any output from our implementation would likely deviate considerably from theirs too as we'd use the whole PolyASite database whereas the authors used 3'seq data from a single study (GEO series GSE30198; SRA accession SRP007359 - ['Defining Cleavage Site Intervals (CSIs)' in Methods section of CSI-UTR paper](https://www.frontiersin.org/articles/10.3389/fgene.2019.00182/full#h3))

As we discussed in the group, we will use their annotation.

**Review by dominikburri — CHANGES_REQUESTED** [review]
Hi @faricazjj thank you for the hard work!

I followed the instructions, downloaded the annotation and sample files and ran the test data. Worked without problems with docker and singularity. 

I agree with @SamBryce-Smith comments, see related posts for requested changes (mainly duplication genes names in bed, identification not valid). 
Also, the files mentioned in `bin/` by Sam should be removed.

**Comment by dominikburri** [comment]
Hi @faricazjj, I ran the workflow also with some human samples and it worked without problems!
I downloaded the annotation and location for Hg38 at the link you provided and adjusted `modules.config` accordingly. 
The resulting differential_out.tsv contains some significant hits (values < 0.05).

**Comment by SamBryce-Smith** [comment]
As I've been digging further into addressing the comments, I'm having a few concerns about which results files we should be using for reporting in the differential output file. 

CSI-UTR has 3 differential expression/usage approaches, all of which are reported in a single run. Previously it was set up to use the 'WITHIN_UTR' results only. I think this should be configurable even if we only end up evaluating one of the modes. [diffUTR evaluated its performance against the DEXSeq mode of CSI-UTR](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04114-7#Sec2), perhaps we could use that as the default.

If you do a test run with our test data (following README instructions as they are) and look at the CSI_OUT/DifferentialExpression` subdirectories, you typically see 2 results files in each subdirectory. The main difference seems to be which condition acts as the denominator and so the direction of the effect size change. I'd expect the p-values to be identical/very similar in this case but wanted to double-check. If identical would mean we can take either of the results files for extracting/reporting p-values.

TL;DR  - WITHIN-UTR & PAIRWISE p-values are identical between the two '.diff.txt' output files. DEXSeq p-values appear to minorly (very minorly) different for the same genes between the different 'denominator' results files.

All below exploring was done using the apaeval_execution_workflows conda environment

## Within-UTR results files

Min p-values for each gene are identical between the two files

```
>>> import pandas as pd
>>> import os
>>> os.getcwd()
'/home/sam/GitHubRepos/APAeval/execution_workflows/CSI-UTR/results/csi_utr/CSI_OUT/DifferentialExpression'
>>> c_vs_s = pd.read_csv('WITHIN_UTR/control_VS_srsf3_CSI.WITHINUTR.diff.txt', sep="\t")
>>> s_vs_c = pd.read_csv('WITHIN_UTR/srsf3_VS_control_CSI.WITHINUTR.diff.txt', sep="\t")
>>> c_vs_s = c_vs_s[['ENSGENE', 'P-value']]
>>> c_vs_s = c_vs_s[['ENSGENE', 'P-value']].groupby('ENSGENE')['P-value'].min().reset_index()
>>> s_vs_c = s_vs_c[['ENSGENE', 'P-value']].groupby('ENSGENE')['P-value'].min().reset_index()
>>> merged = pd.merge(c_vs_s, s_vs_c, how='left', on='ENSGENE')
>>> merged.head()
              ENSGENE  P-value_x  P-value_y
0  ENSMUSG00000000001   0.999999   0.999999
1  ENSMUSG00000000003   1.000000   1.000000
2  ENSMUSG00000000028   1.000000   1.000000
3  ENSMUSG00000000049   1.000000   1.000000
4  ENSMUSG00000000056   0.999999   0.999999
>>> (merged['P-value_x'] == merged['P-value_y']).all()
True

```

## Pairwise results files

Again all P-values are identical. Don't know if can conclude fully as all p-values are 1 for this approach.

```
>>> import pandas as pd
>>> import os
>>> os.getcwd()
'/home/sam/GitHubRepos/APAeval/execution_workflows/CSI-UTR/results/csi_utr/CSI_OUT/DifferentialExpression'
>>> c_vs_s = pd.read_csv('PAIRWISE/control_VS_srsf3_PAIRWISE_CSI.diff.txt', sep="\t")
>>> s_vs_c = pd.read_csv('PAIRWISE/srsf3_VS_control_PAIRWISE_CSI.diff.txt', sep="\t")
>>> c_vs_s = c_vs_s[['ENSEMBL_GENE_ID', 'P-Value']].groupby('ENSEMBL_GENE_ID')['P-Value'].min().reset_index()
>>> s_vs_c = s_vs_c[['ENSEMBL_GENE_ID', 'P-Value']].groupby('ENSEMBL_GENE_ID')['P-Value'].min().reset_index()
>>> merged = pd.merge(c_vs_s, s_vs_c, how='left', on='ENSEMBL_GENE_ID')
>>> merged.head()
      ENSEMBL_GENE_ID  P-Value_x  P-Value_y
0  ENSMUSG00000000001          1          1
1  ENSMUSG00000000003          1          1
2  ENSMUSG00000000028          1          1
3  ENSMUSG00000000049          1          1
4  ENSMUSG00000000056          1          1
>>> (merged['P-Value_x'] == merged['P-Value_y']).all()
True
>>> merged.loc[merged['P-Value_x'] < 1, :]
Empty DataFrame
Columns: [ENSEMBL_GENE_ID, P-Value_x, P-Value_y]
Index: []

```

## DEXSeq results files

Big difference here with test data is that most of the genes have an NA p-value (probably due to zero counts). I don't think we've decided how to handle these, for now I've reported them in the output TSV with 'NA' in the 2nd column but can discuss that later.

P-values between the different 'denominator files' are not identical, but appear to be very very similar.

```
(apaeval_execution_workflows) sam@sam-XPS-13-7390:~/GitHubRepos/APAeval/execution_workflows/CSI-UTR/results/csi_utr/CSI_OUT/DifferentialExpression$ python
Python 3.9.4 | packaged by conda-forge | (default, May 10 2021, 22:13:33) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pandas as pd
>>> import os
>>> os.getcwd()
'/home/sam/GitHubRepos/APAeval/execution_workflows/CSI-UTR/results/csi_utr/CSI_OUT/DifferentialExpression'
>>> c_vs_s = pd.read_csv('DEXSeq/control_VS_srsf3/allCSIs_control_VS_srsf3_BYREGION.txt', sep="\t")
>>> c_vs_s.head()
                                       CSI_ID             GENE_ID  ... countData.siControl_R2_2genes           transcripts
0  ENSMUSG00000025902:4491716_4490930-4489289  ENSMUSG00000025902  ...                             0  ENSMUSG00000025902_1
1  ENSMUSG00000025902:4491716_4491389-4490930  ENSMUSG00000025902  ...                             0  ENSMUSG00000025902_2
2  ENSMUSG00000025902:4491716_4491716-4491389  ENSMUSG00000025902  ...                             0  ENSMUSG00000025902_3
3  ENSMUSG00000033845:4776464_4774300-4773211  ENSMUSG00000033845  ...                             0  ENSMUSG00000033845_1
4  ENSMUSG00000033845:4776464_4774436-4774300  ENSMUSG00000033845  ...                             0  ENSMUSG00000033845_2

[5 rows x 24 columns]
>>> s_vs_c = pd.read_csv('DEXSeq/srsf3_VS_control/allCSIs_srsf3_VS_control_BYREGION.txt', sep="\t")
>>> c_vs_s = c_vs_s[['GENE_ID', 'pvalue']]
>>> s_vs_c = s_vs_c[['GENE_ID', 'pvalue']]
>>> c_vs_s = c_vs_s[['GENE_ID', 'pvalue']].groupby('GENE_ID')['pvalue'].min().reset_index()
>>> s_vs_c = s_vs_c[['GENE_ID', 'pvalue']].groupby('GENE_ID')['pvalue'].min().reset_index()
>>> merged = pd.merge(c_vs_s, s_vs_c, how='left', on='GENE_ID')
>>> merged.head()
              GENE_ID  pvalue_x  pvalue_y
0  ENSMUSG00000000001       NaN       NaN
1  ENSMUSG00000000003       NaN       NaN
2  ENSMUSG00000000028       NaN       NaN
3  ENSMUSG00000000049       NaN       NaN
4  ENSMUSG00000000056       NaN       NaN
>>> (merged['pvalue_x'] == merged['pvalue_y']).all()
False
>>> (merged['pvalue_x'].dropna() == merged['pvalue_y'].dropna()).all()
False
>>> (merged['pvalue_x'].dropna() == merged['pvalue_y'].dropna())
3610     False
3612     False
11035     True
12037    False
dtype: bool
>>> merged.dropna()
                  GENE_ID      pvalue_x      pvalue_y
3610   ENSMUSG00000022863  4.498875e-01  4.498876e-01
3612   ENSMUSG00000022865  1.053433e-45  1.053433e-45
11035  ENSMUSG00000041372  1.000000e+00  1.000000e+00
12037  ENSMUSG00000045962  2.114934e-91  2.114934e-91

>>> 
```

Will commit changes soon but wanted to be able to refer to this comment in the code/README

**Review by yuukiiwa — APPROVED** [review]
Hi @faricazjj, @dominikburri, and @SamBryce-Smith, this looks really good to me! 
Thank you so much for the hard work! :)

**Review by dominikburri — APPROVED** [review]
Hi @SamBryce-Smith, thank you for the speedy changes!

And great that you spotted the different differential results output. Setting DEXSeq as default is great, as we try to always use the default parameters. 

I ran the test data without problems. 

Good to be merged from my side.

File: execution_workflows/CSI-UTR/bin/postprocess_differential.py [hunk 2]
```diff
@@ -1,59 +1,107 @@
...
-    differential_out.close()
+    # Find results files for specified result type to reprt
+    if result_type == "DEXSeq":
+        # In DEXSeq results dir, results tables are stored under subdirs (comparison direction)
+        subdirs = [join(csi_utr_results_dir, f) for f in listdir(csi_utr_results_dir) if isdir(join(csi_utr_results_dir, f))]
+
+        assert len(subdirs) == 2, f"Expect 2 'DEXSeq' subdirectories within provided directory, {', '.join(subdirs)} were found"
+
+        sys.stderr.write(f"DEXSeq subdirectories found - {', '.join(subdirs)}\n")
+        sys.stderr.write(f"P-values are the same since only effect size direction is changed - choosing {subdirs[0]}\n")
```
[the remark this answers is not in the document]
Comment (SamBryce-Smith) [reply execution_workflows/CSI-UTR/bin/postprocess_differential.py:67]: Good point definitely just a code habit on my part. Very occasionally I have scripts outputting to stdout but still want to track/have some info/log messages. In those cases my dirty workaround is just to print the log messages to stderr and redirect.
